src/serverd.cpp

```diff
@@ -0,0 +1,184 @@
+//===--- serverd.cpp - Tasknow daemon socket server -----------------------===//
+//
+// Copyright (c) 2023 Yuri Istomin
+//
+// Part of the NekoCdr's "tasknow" project
+// Licensed under the Apache License, Version 2.0 (the "License").
+// See http://www.apache.org/licenses/LICENSE-2.0 for license information.
+//
+//===----------------------------------------------------------------------===//
+
+#include "serverd.h"
+
+#include "defines.h"
+
+#include <cerrno>
+#include <cstdio>
+#include <cstring>
+#include <format>
+#include <iostream>
+#include <string_view>
+#include <sys/socket.h>
+#include <sys/un.h>
+#include <unistd.h>
+
+namespace tasknow::serverd {
+
+auto init(
+    struct sockaddr_un* server_addr,
+    struct sockaddr_un* client_addr,
+    socklen_t addr_len,
+    const char* sock_path
+) -> void {
+    memset(server_addr, 0, addr_len);
+    server_addr->sun_family = AF_UNIX;\
+    // TODO: add check for sock_path length
+    strcpy(static_cast<char*>(server_addr->sun_path), sock_path);
+
+    memset(client_addr, 0, addr_len);
+
+    if (std::remove(sock_path) == ErrorCode && errno != ENOENT) {
+        throw std::string_view{"Can't delete socket file"};
+    }
+}
+
+auto create_socket(int* server_sock) -> void
+{
+    *server_sock = socket(AF_UNIX, SOCK_STREAM, 0);
+    if (*server_sock == ErrorCode) {
+        throw std::string_view{std::format("Socket error: %d", errno)};
+    }
+}
+
+auto bind_socket(
+    int* server_sock,
+    struct sockaddr_un* server_addr,
+    socklen_t addr_len
+) -> void {
+    if (bind(
+        *server_sock,
+        // NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
+        reinterpret_cast<struct sockaddr*>(server_addr),
+        addr_len
+    ) == ErrorCode) {
+        throw std::string_view{std::format("Bind socket error: %d", errno)};
+    }
+}
+
+auto listen_socket(int* server_sock, const int backlog_size) -> void
+{
+    if (listen(*server_sock, backlog_size) == ErrorCode) {
+        throw std::string_view{std::format("Listen socket error: %d", errno)};
+    }
+}
+
+auto accept_client(
+    int* server_sock,
+    int* client_sock,
+    struct sockaddr_un* client_addr,
+    socklen_t* addr_len
+) -> void {
+    *client_sock = accept(
+        *server_sock,
+        // NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
+        reinterpret_cast<struct sockaddr*>(client_addr),
+        addr_len
+    );
+    if (*client_sock == ErrorCode) {
+        throw std::string_view{"Can't accept client"};
```

Comment:
If you add `using namespace std::literals;` to your headers, you'd be able to write this much shorter: `throw "Can't accept client"sv;`